Allow Access Token and OIDC Client scope control

bigkevmcd opened the pull request:
## Issue: https://github.com/rancher/rancher/issues/52716
<!-- If your PR depends on changes from another pr link them here and describe why they are needed on your solution section. -->
 
## Problem
<!-- Describe the root cause of the issue you are resolving. This may include what behavior is observed and why it is not desirable. If this is a new feature describe why we need this feature and how it will be used. -->
The Access token that is issued by the OIDC Provider can't be used to authenticate to Rancher.

## Solution
<!-- Describe what you changed to fix the issue. Relate your changes back to the original issue / feature and explain why this addresses the issue. -->

This adds two new features...
 * Support for configuring the allowed scopes that a relying party can use with a set of client credentials
 * The Access Token from the OIDC Exchange in the OIDC Provider can now be used with the Rancher API
 
The allowed scopes are not currently used beyond validating that the authorize request scopes are allowed by the client.
If an OIDC Client has no configured scopes, it will default to the previously supported scopes i.e. `openid`, `offline_access` and `profile`.

## Testing
<!-- Note: Confirm if the repro steps in the GitHub issue are valid, if not, please update the issue with accurate repro steps. -->
Create an `OIDCClient` with or without scopes, authenticate via OIDC and use the OIDC Access Token to access the Rancher API.

Adding scopes and changing the client to request those scopes will result in errors from the API.

## Engineering Testing
### Manual Testing
<!-- Describe what manual testing you did (if no testing was done, explain why). -->
Created a small Go tool that bounces through Rancher for OIDC and then queries the API with the token and this works.

I'm happy to make this (and the `OIDCClient`) available 😄 

### Automated Testing
<!-- Ensure there are unit/integration/validation tests added (if possible); describe what cases they cover and do not cover. -->
* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  <!-- 
  Non-exhaustive list of reasons:
    - Lack of the framework capable of testing this fix/change
    - Tight deadlines / critical priority to get fix/change in - !ensure GH issue is logged to add tests!
    - No application logic is modified by this change, e.g. refactoring/cosmetic/non-code/test change
    - Tests implemented in another PR elsewhere - !ensure GH PR link is added!
    - Other (explain)
  Note: Outside of the exceptions above, the "existing tests cover the changes" is very unlikely to be an acceptable reason as the existing tests generally don't cover the logic changes implemented by this PR 
  -->
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

## QA Testing Considerations
<!-- Highlight areas or (additional) cases that QA should test w.r.t a fresh install as well as the upgrade scenarios -->
 
### Regressions Considerations
<!-- Dedicated section to specifically call out any areas that with higher chance of regressions caused by this change, include estimation of probability of regressions -->
_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

bigkevmcd committed:
Add support OIDC-Provider issued access tokens.

This adds functionality for using an OIDC Provider issued access token
to authenticate with the Rancher API.

bigkevmcd committed:
Implement scope checking.

OIDCClients can be configured to only allow specific scopes.

raulcabello commented on pkg/oidc/provider/token.go:
```diff
@@ -408,6 +375,52 @@ func (h *tokenHandler) createTokenResponse(rancherToken *v3.Token, oidcClient *v
+		"iat":   now.Unix(),
+		"sub":   rancherToken.UserID,
+		"scope": scopes,
+		"token": rancherToken.Name + ":" + rancherToken.Token,
```
we might want to consider adding only the `rancherToken.Name` here as the `rancherToken` probably expires after the `access_token`. However, that would mean and extra call to the cache to get the token value when authorizing. What do you think?

raulcabello reviewed (commented):
Overall this looks great! I tested using an access token in a kubeconfig and it worked as expected. I have left a few comments.

bigkevmcd committed:
Split out createIDToken from the token handler.

Co-authored-by: Raul Cabello Martin <raulcabm@gmail.com>

bigkevmcd committed:
Do not check for openid scope.

bigkevmcd committed:
Remove the token from the AccessToken.

bigkevmcd committed:
Don't return ID token if no openid scope.

bigkevmcd committed:
Only enable OIDC auth if OIDC Feature enabled.

bigkevmcd committed:
Additional tests for OIDC token auth.

Some lint cleanup of the authentication tests.

pmatseykanets reviewed (commented):
Looks good! A minor concern with type assertion and unresolved questions from Raul.

bigkevmcd committed:
Harden token parsing around the kid field

This ensures that the "kid" in the token is parsed as a string correctly
and will error if it's not a valid string.

This prevents signed tokens with a non-string field from causing a
panic.

The documentation is clarified.
